rust/kernel/types.rs

```diff
@@ -463,3 +463,102 @@ impl<T: LittleEndian + Copy> core::convert::From<T> for LE<T> {
         LE(value.to_le())
     }
 }
+
+/// Specifies that a type is safely readable from byte slices.
+///
+/// Not all types can be safely read from byte slices; examples from
+/// <https://doc.rust-lang.org/reference/behavior-considered-undefined.html> include [`bool`] that
+/// must be either `0` or `1`, and [`char`] that cannot be a surrogate or above [`char::MAX`].
+///
+/// # Safety
+///
+/// Implementers must ensure that the type is made up only of types that can be safely read from
+/// arbitrary byte sequences (e.g., [`u32`], [`u64`], etc.).
+pub unsafe trait FromBytes: Sized {
+    /// Converts the given byte slice into a shared reference to [`Self`].
+    ///
+    /// It fails if the size or alignment requirements are not satisfied.
+    fn from_bytes(data: &[u8], offset: usize) -> Option<&Self> {
```

Comment:
> Why is there an `offset` argument to this function? When calling this function, I could just do

Because all uses of this function (introduced in tarfs) have an offset.

> 
> ```rust
> T::from_bytes(&data[offset..])
> ```

These are not equivalent when `offset` is out of bounds. This version returns `None`, yours panics.

> If you really want to keep it, I would suggest this to be the first line of the function:
> 
> ```rust
> let data = &data[offset];
> ```
> 
> From that point onward you do not need to consider `offset`.

Do you mean `&data[offset..]`? Yes, I think that's better. Made the change.

It also occurs to me that `&data[0]` doesn't work if the offset is the length of the slice, so changing it to `data.as_ptr()`.

Reply:
> This version returns None, yours panics.
Oh I see, that is unfortunate. Then this approach is good.